JOSESwift/Sources/CryptoImplementation/EC.swift

```diff
@@ -175,25 +175,32 @@ internal struct EC {
     /// - Returns: The signature.
     /// - Throws: `ECError` if any errors occur while signing the input data.
     static func sign(_ signingInput: Data, with privateKey: KeyType, and algorithm: SignatureAlgorithm) throws -> Data {
-        // Sign the input as raw elliptic curve coordinates using a hashing algorithm and a private key.
         guard let curveType = algorithm.curveType else {
             throw ECError.invalidCurveDigestAlgorithm
         }
-
-        let digest = try algorithm.createDigest(input: signingInput)
-        var signatureLength = curveType.signatureOctetLength
-
-        guard let signature = NSMutableData(length: signatureLength) else {
-            throw ECError.couldNotAllocateMemoryForSignature
+        guard let secKeyAlgorithm = algorithm.secKeyAlgorithm else {
+            throw ECError.algorithmNotSupported
         }
 
-        let signatureBytes = signature.mutableBytes.assumingMemoryBound(to: UInt8.self)
-        let status = SecKeyRawSign(privateKey, .sigRaw, digest, digest.count, signatureBytes, &signatureLength)
-        if status != errSecSuccess {
-            throw ECError.signingFailed(description: "Error creating signature. (OSStatus: \(status))")
+        var cfErrorRef: Unmanaged<CFError>?
+        let signature = SecKeyCreateSignature(privateKey, secKeyAlgorithm, signingInput as CFData, &cfErrorRef)
+        if cfErrorRef != nil {
+            throw ECError.signingFailed(description: "Error creating signature. (CFError: \(cfErrorRef!.takeRetainedValue()))")
         }
 
-        return signature as Data
+        // unpack BER encoded ASN.1 format signature to raw format as specified for JWS
+        let ecSignatureTLV = [UInt8](signature! as Data)
+        do {
+            let ecSignature = try ecSignatureTLV.read(.sequence)
+            let varlenR = try Data(ecSignature.read(.integer))
+            let varlenS = try Data(ecSignature.skip(.integer).read(.integer))
+            let fixlenR = varlenR.withLengthFixedTo(curveType.coordinateOctetLength)
+            let fixlenS = varlenS.withLengthFixedTo(curveType.coordinateOctetLength)
+
+            return fixlenR + fixlenS
```

Comment:
> One more suggestion from my side: What would you think about moving the DER/ASN.1 packing and unpacking code to functions that can be tested independently from the signing/verifying code
> logic? Since this is the crucial feature of this pull request, it would be awesome to be able to test them independently.

I think testing this logic independently would be a good idea, but I would not want to extend the implementation to support parsing more than what is strictly required for this use case or have these functions exposed to other parts of the code in a way that suggests they can be reused elsewhere when actually right now they only work within the constraints that apply for this use case.

I have to look into how to scope things in Swift to achieve this and have it be testable.

Reply:
That is a good point. For testing, we would have to raise the access control level from `fileprivate` to `internal`. I totally agree with you that that might suggest that the functions could be used from elsewhere in the code base- That's certainly not what we want.

What would you think of having them as an extension of `ECCurveType` that would clearly namespace it as something `EC` internal and would even give you direct access to the `coordinateOctetLength`. Just an idea. Let me know what you think, I'm open for suggestions.

Reply:
I am halfway-done with the tests on my work machine. Currently I am using another struct inside of the `EC` struct with two static functions to avoid the extension getting exposed everywhere, what you are suggesting sounds like it may be a better idea.

Reply:
I'll leave it up to you how to best handle this. 👍